tests/test_pin_lock_advanced.py

```diff
@@ -0,0 +1,220 @@
+        with mock.patch.object(_QuadPackageDef, 'allocate', autospec=True) as mock_allocate:
+            # Configure the mock to return predictable values
+            mock_allocate.return_value = ["10", "11"]
+
+            # Set up package definitions with our real Pydantic object
+            mock_package_defs.__getitem__.return_value = package_def
+            mock_package_defs.__contains__.return_value = True
+
+            # Create real Pydantic objects for configuration instead of mocks
+            # Start with pads and power
+            pads = {}
+            for name, config in self.mock_config["chipflow"]["silicon"]["pads"].items():
+                pads[name] = PadConfig(
+                    type=config["type"],
+                    loc=config["loc"]
+                )
+
+            power = {}
+            for name, config in self.mock_config["chipflow"]["silicon"]["power"].items():
+                power[name] = PadConfig(
+                    type=config["type"],
+                    loc=config["loc"]
+                )
+
+            # Create a Silicon config object
+            silicon_config = SiliconConfig(
+                process="ihp_sg13g2",
+                package="pga144",
+                pads=pads,
+                power=power
+            )
+
+            # Create the Chipflow config object with proper StepsConfig
+            from chipflow_lib.config_models import StepsConfig
+
+            steps_config = StepsConfig(
+                silicon="chipflow_lib.steps.silicon:SiliconStep"
+            )
+
+            chipflow_config = ChipFlowConfig(
+                project_name="test_project",
+                silicon=silicon_config,
+                steps=steps_config
+            )
+
+            # Create the full Config object
+            config_model = Config(chipflow=chipflow_config)
+
+            # Set up the mock model_validate to return our real Pydantic config object
+            mock_model_validate.return_value = config_model
+
+            # Set up parse_config to return the dict version
+            mock_parse_config.return_value = self.mock_config
+
+            # Mock top interfaces to return something simple
+            mock_interface = {
+                "component1": {
+                    "interface": {
+                        "members": {
+                            "uart": {
+                                "type": "interface",
+                                "members": {
+                                    "tx": {"type": "port", "width": 1, "dir": "o"},
+                                    "rx": {"type": "port", "width": 1, "dir": "i"}
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            mock_top_interfaces.return_value = ({}, mock_interface)
+
+            # Call lock_pins with mocked file operations
+            with mock.patch('builtins.print'), \
```

Comment:
why is `builtins.print` being mock-patched here?

Reply:
good catch, a left-over